File: github/merge.py
"""Merge a GitHub pull request via the gh CLI."""

from __future__ import annotations

import subprocess
from typing import Any

VALID_STRATEGIES = {"merge", "squash", "rebase"}


def merge_pr(
    repo: str,
    pr_number: int,
    strategy: str = "merge",
) -> dict[str, Any]:
    """Merge a pull request using ``gh pr merge``.

    Args:
        repo: Repository in ``owner/repo`` format.
        pr_number: The PR number to merge.
        strategy: One of ``merge``, ``squash``, or ``rebase``.

    Returns:
        A dict with ``success`` (bool), ``repo``, ``pr_number``, and
        ``error`` (str) on failure.
    """
    if strategy not in VALID_STRATEGIES:
        raise ValueError(f"strategy must be one of {sorted(VALID_STRATEGIES)}, got: {strategy!r}")

    cmd = ["gh", "pr", "merge", str(pr_number), "--repo", repo, f"--{strategy}"]
    result = subprocess.run(cmd, capture_output=True, text=True)

    if result.returncode != 0:
        return {
            "success": False,
            "repo": repo,
            "pr_number": pr_number,
            "error": result.stderr.strip(),
        }

    return {
        "success": True,
        "repo": repo,
        "pr_number": pr_number,
    }

File: internal/github/merge.go
package github

import (
	"context"
	"fmt"
	"os/exec"
)

// MergeResult represents the outcome of a PR merge attempt.
type MergeResult struct {
	Success  bool   `json:"success"`
	Repo     string `json:"repo"`
	PRNumber int    `json:"pr_number"`
	Error    string `json:"error,omitempty"`
}

// MergePR merges a PR using the gh CLI.
// Strategy can be "merge", "squash", or "rebase".
func MergePR(ctx context.Context, repo string, prNumber int, strategy string) MergeResult {
	if strategy == "" {
		strategy = "merge"
	}

	args := []string{"pr", "merge", fmt.Sprintf("%d", prNumber),
		"--repo", repo, "--" + strategy,
	}

	cmd := exec.CommandContext(ctx, "gh", args...)
	if out, err := cmd.CombinedOutput(); err != nil {
		return MergeResult{
			Success:  false,
			Repo:     repo,
			PRNumber: prNumber,
			Error:    fmt.Sprintf("%s: %s", err, string(out)),
		}
	}

	return MergeResult{
		Success:  true,
		Repo:     repo,
		PRNumber: prNumber,
	}
}
